src/spec/file.spec.js

```diff
@@ -53,4 +70,337 @@ describe('File', () => {
+
+      source.on('error', (error) => {
+        expect(error).toBeInstanceOf(Error);
+
+        file.handle = originalHandle;
+        file.close();
+        mpq.close();
+        done();
+      });
+
+      file.handle = invalidHandle;
+
+      source.pipe(destination);
+    });
+
+    test('emits error when setting pos in piping file with invalid handle', async (done) => {
+      expect.assertions(1);
+
+      const mpq = await MPQ.open('/fixture/vanilla-standard.mpq');
+      const file = mpq.openFile('fixture.txt');
+
+      const source = file.createStream();
+      const destination = new WritableMemoryStream();
+
+      const originalHandle = file.handle;
+      const invalidHandle = new StormLib.VoidPtr();
+
+      source.on('error', (error) => {
+        expect(error).toBeInstanceOf(Error);
+
+        file.handle = originalHandle;
+        file.close();
+        mpq.close();
+        done();
+      });
+
+      file.handle = invalidHandle;
+
+      source.pipe(destination);
+    });
+
+    test('reads valid file', async () => {
+      const mpq = await MPQ.open('/fixture/vanilla-standard.mpq');
+      const file = mpq.openFile('fixture.txt');
+
+      const result = file.read();
+
+      expect(result).toEqual(new Uint8Array([
+        102, 105, 120, 116, 117, 114, 101, 45, 116, 101, 120, 116, 10
+      ]));
+
+      file.close();
+      mpq.close();
+    });
+
+    test('reads valid file twice', async () => {
+      const mpq = await MPQ.open('/fixture/vanilla-standard.mpq');
+      const file = mpq.openFile('fixture.txt');
+
+      const read1 = file.read();
+      const read2 = file.read();
+
+      expect(read1).toEqual(new Uint8Array([
+        102, 105, 120, 116, 117, 114, 101, 45, 116, 101, 120, 116, 10
+      ]));
+
+      expect(read2).toEqual(read1);
+
+      file.close();
+      mpq.close();
+    });
+
+    test('throws if reading closed file', async () => {
+      expect.assertions(1);
+
+      const mpq = await MPQ.open('/fixture/vanilla-standard.mpq');
+      const file = mpq.openFile('fixture.txt');
+      file.close();
+
+      try {
+        const data = file.read();
+      } catch (err) {
+        expect(err).toBeInstanceOf(Error);
+      } finally {
+        mpq.close();
+      }
+    });
+
+    test('throws if reading file with invalid handle', async () => {
+      expect.assertions(1);
+
+      const mpq = await MPQ.open('/fixture/vanilla-standard.mpq');
+      const file = mpq.openFile('fixture.txt');
+
+      Object.defineProperty(file, 'pos', { set: () => {} });
+      Object.defineProperty(file, 'size', { get: () => 1 });
+
+      const originalHandle = file.handle;
+      const invalidHandle = new StormLib.VoidPtr();
+
+      file.handle = invalidHandle;
+
+      try {
+        const data = file.read();
+      } catch (err) {
+        expect(err).toBeInstanceOf(Error);
+      } finally {
+        file.handle = originalHandle;
+        file.close();
+        mpq.close();
+      }
```

Comment:
We could potentially use `.toThrow()` here:

```javascript
expect(() => {
  file.read();
}).toThrow(Error, 'Invalid handle');
```

Reply:
👍 I'll update to use `toThrow`.

To reduce coupling, I'm leaning toward not including the error string for now.

I know it's overly broad, and thus could produce misleading test results. In a future PR, I'd propose we tackle that by introducing a handful of `Error` subclasses specific to various categories of StormJS errors, and update the test suites to compare with the subclasses.

Thoughts?

Reply:
Agree, sounds great 👍